stats/opentelemetry/csm/pluginoption.go

```diff
@@ -0,0 +1,336 @@
+// metadata exchange labels to be sent as part of metadata exchange from a CSM
+// Plugin Option.
+func initializeLocalAndMetadataLabels(labels map[string]string) (map[string]string, string) {
+	// The value of “csm.workload_canonical_service” comes from
+	// “CSM_CANONICAL_SERVICE_NAME” env var, “unknown” if unset.
+	val := labels["canonical_service"]
+	localLabels := make(map[string]string)
+	localLabels["csm.workload_canonical_service"] = val
+	// Get the CSM Mesh ID from the bootstrap file.
+	nodeID := getNodeID()
+	localLabels["csm.mesh_id"] = parseMeshIDFromNodeID(nodeID)
+
+	// Metadata exchange labels - can go ahead and encode into proto, and then
+	// base64.
+	pbLabels := &structpb.Struct{
+		Fields: map[string]*structpb.Value{},
+	}
+	for k, v := range labels {
+		pbLabels.Fields[k] = structpb.NewStringValue(v)
+	}
+	protoWireFormat, err := proto.Marshal(pbLabels)
+	metadataExchangeLabelsEncoded := ""
+	if err == nil {
+		metadataExchangeLabelsEncoded = base64.RawStdEncoding.EncodeToString(protoWireFormat)
+	}
+	// else - This behavior triggers server side to reply (if sent from a gRPC
+	// Client within this binary) with the metadata exchange labels. Even if
+	// client side has a problem marshaling proto into wire format, it can
+	// still use server labels so send an empty string as the value of
+	// x-envoy-peer-metadata. The presence of this metadata exchange header
+	// will cause server side to respond with metadata exchange labels.
+
+	return localLabels, metadataExchangeLabelsEncoded
+}
+
+// getNodeID gets the Node ID from the bootstrap data.
+func getNodeID() string {
+	cfg, err := bootstrap.NewConfig()
+	if err != nil {
+		return "" // will become "unknown"
+	}
+	if cfg.NodeProto == nil {
+		return ""
+	}
+	return cfg.NodeProto.GetId()
+}
+
+// metadataExchangeKey is the key for HTTP metadata exchange.
+const metadataExchangeKey = "x-envoy-peer-metadata"
+
+func determineTargetCSM(target string) bool {
+	// On the client-side, the channel target is used to determine if a channel is a
+	// CSM channel or not. CSM channels need to have an “xds” scheme and a
+	// "traffic-director-global.xds.googleapis.com" authority. In the cases where no
+	// authority is mentioned, the authority is assumed to be CSM. MetadataExchange
+	// is performed only for CSM channels. Non-metadata exchange labels are detected
+	// as described below.
+	parsedTarget, err := url.Parse(target)
+	if err != nil {
+		// Shouldn't happen as Dial would fail if target couldn't be parsed, but
+		// log just in case to inform user.
+		logger.Errorf("passed in target %v failed to parse: %v", parsedTarget, err)
+		return false
+	}
+
+	if parsedTarget.Scheme == "xds" {
+		if parsedTarget.Host == "" {
+			return true // "In the cases where no authority is mentioned, the authority is assumed to be csm"
+		}
+		return parsedTarget.Host == "traffic-director-global.xds.googleapis.com"
+	}
+	return false
```

Comment:
Optional: FWIW this can be simplified:

```go
return parsedTarget.Scheme == "xds" && (parsedTarget.Host == "" || parsedTarget.Host == tdGlobal)
```

Reply:
Oh nice one.